Source/santad/SNTApplicationTest.m

```diff
@@ -0,0 +1,129 @@
+@property id mockSNTDatabaseController;
+@end
+
+@implementation SNTApplicationTest
+- (void)setUp {
+  [super setUp];
+  fclose(stdout);
+  self.mockSNTDatabaseController = OCMClassMock([SNTDatabaseController class]);
+  XCTAssertTrue([[SNTConfigurator configurator] enableSystemExtension]);
+}
+
+- (void)tearDown {
+  [self.mockSNTDatabaseController stopMocking];
+  [super tearDown];
+}
+
+- (void)checkBinaryExecution:(NSString *)binaryName
+                    testPath:(NSString *)testPath
+                  wantResult:(es_auth_result_t)wantResult {
+  MockEndpointSecurity *mockES = [MockEndpointSecurity mockEndpointSecurity];
+  [mockES reset];
+
+  OCMStub([self.mockSNTDatabaseController databasePath]).andReturn(testPath);
+
+  SNTApplication *app = [[SNTApplication alloc] init];
+  [app start];
+
+  // es events will start flowing in as soon as es_subscribe is called, regardless
+  // of whether we're ready or not for it.
+  while (!mockES.subscribed)
+    ;
+
+  XCTestExpectation *expectation =
+    [self expectationWithDescription:@"Wait for santa's Auth dispatch queue"];
+  __block ESResponse *got = nil;
+  [mockES registerResponseCallback:^(ESResponse *r) {
+    @synchronized(self) {
+      got = r;
+      [expectation fulfill];
+    }
+  }];
+
+  NSString *binaryPath = [NSString pathWithComponents:@[ testPath, binaryName ]];
+  es_file_t binary = {.path = MakeStringToken(binaryPath)};
+  es_process_t proc = {
+    .executable = &binary,
+    .is_es_client = false,
+    .codesigning_flags = 570509313,
+    .session_id = 12345,
+    .group_id = 12345,
+    .ppid = 12345,
+    .original_ppid = 12345,
+    .is_platform_binary = false,
+  };
+  es_event_exec_t exec_event = {
+    .target = &proc,
+  };
+  es_events_t event = {.exec = exec_event};
+  es_message_t m = {
+    .version = 4,
+    .mach_time = 181576143417379,
+    .event_type = ES_EVENT_TYPE_AUTH_EXEC,
+    .event = event,
+    .action_type = ES_ACTION_TYPE_AUTH,
+    .deadline = DISPATCH_TIME_FOREVER,
+    .process = &proc,
+    .seq_num = 1,
+  };
+
+  [mockES triggerHandler:&m];
+
+  [self
+    waitForExpectationsWithTimeout:5.0
+                           handler:^(NSError *error) {
+                             if (error) {
+                               XCTFail(
+                                 @"Santa auth test on binary \"%@/%@\" timed out with error: %@",
+                                 testPath, binaryName, error);
+                             }
+                           }];
+
+  XCTAssertEqual(got.result, wantResult, @"received unexpected ES response on executing \"%@/%@\"",
+                 testPath, binaryName);
+}
+
+- (void)testBinaryRules {
+  NSString *testPath = @"santa/Source/santad/testdata/binaryrules";
+  NSDictionary *testCases = @{
+    @"badbinary" : [NSNumber numberWithInt:ES_AUTH_RESULT_DENY],
+    @"goodbinary" : [NSNumber numberWithInt:ES_AUTH_RESULT_ALLOW],
+  };
+  NSString *fullTestPath = [NSString pathWithComponents:@[
```

Comment:
If you switch from data -> resources in the build file, this would be:

```
NSString *path = [[NSBundle bundleForClass:[self class]] pathForResource:@"binaryrules" ofType:@""];
```